### DISCUSSION TITLE
`自定义顶栏`中历史、收藏和稍后再看组件显示分 P 数、视频时长、观看进度相关建议

### DISCUSSION THREAD

**Discussion by Geeyun-JY3** [the1812/Bilibili-Evolved#1866] [功能建议 / Ideas]
（以下实例配图中视频均为多 P 视频）

### 现状省流表格（实测更适合 PC 网页端查看）：

|                       | 历史（记录） | 收藏 | 稍后再看 |
| --------------------- | ---------- | ---- | ------ |
| 自定义顶栏（f53902062） | <li>最近观看分 P 的时长</li><li>最近观看分 P 的观看进度百分比</li><li>进度条</li>![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/8204d9b2-a603-4d58-8309-e3743fbf9761) | 所有分 P 的总时长<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/7b6c2125-0a19-4b95-b464-3e404b84e60b) | 所有分 P 的总时长<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/ea4755cb-cf17-4241-9ce1-6ffafc0d215a) |
| 旧版原版顶栏 | 仅进度条<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/38b0ae35-b5f6-416c-aec0-a0cd86b8f00a) | 分 P 数<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/3b6b90b6-e45b-4d00-97af-89f2ed6a1db4) | 分 P 数、时长、进度均无<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/3761be8e-5a75-4226-8975-ee3b7316bcd1) |
| 新版原版顶栏 | <li>最近观看分 P 的时长</li><li>最近观看分 P 的具体观看进度</li><li>进度条</li>![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/1cbc82c7-0b39-4645-8da4-10e7c9dfa1d3) | 所有分 P 的总时长<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/45d0fb4b-aeb9-4fb8-aca3-00663a5b0b2a) | 所有分 P 的总时长<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/4c5e974b-a544-4bc0-957f-669264c36483) |
| Android 客户端[^1] | <li>最近观看分 P 的时长</li><li>最近观看分 P 的具体观看进度</li><li>进度条</li><li>最近观看分 P 名（同时不显示视频所属 UP 主）</li>![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/b8ff8b91-11ee-4f21-90cf-481b63d9a5ca) | 分 P 数<br />![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/2ecc01b4-40cf-4933-b62c-5c653f9d502e) | <li>所有分 P 的总时长</li><li>所有分 P 的具体观看进度</li><li>进度条</li>![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/540e610a-6927-41b5-83f3-7f13e25fa6be) |

[^1]: Play 版，版本号 3.19.0 (7750300)

### 目前`自定义顶栏`中：

- `历史`组件只会显示最近观看分 P 的时长，且光标悬停时封面左上角会显示已观看百分比（观看进度），封面下方有一进度条指示大致观看进度：
![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/8204d9b2-a603-4d58-8309-e3743fbf9761)
- 而`收藏`和`稍后再看`组件会显示所有分 P 的总时长：
  - `收藏`：
  ![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/7b6c2125-0a19-4b95-b464-3e404b84e60b)
  - `稍后再看`：
  ![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/ea4755cb-cf17-4241-9ce1-6ffafc0d215a)

### 而原版顶栏中：

- `收藏`只会显示分 P 数：
![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/3b6b90b6-e45b-4d00-97af-89f2ed6a1db4)
- `稍后再看`和`历史`则是时长、观看进度和分 P 数都不会显示：
  - `历史`（封面下方有一进度条指示大致观看进度）：
  ![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/38b0ae35-b5f6-416c-aec0-a0cd86b8f00a)
  - `稍后再看`：
  ![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/3761be8e-5a75-4226-8975-ee3b7316bcd1)

***

2024-06-04 编辑：

### 众所周知新版页面的原版顶栏也改过了，现在时长显示相关变成了这样：

- `历史`只会显示最近观看分 P 的时长，且封面右下角会显示最近观看分 P 的具体观看进度（形如`具体看到的时间点/最近观看分 P 的时长`），封面下方也还有一进度条指示大致观看进度：
![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/1cbc82c7-0b39-4645-8da4-10e7c9dfa1d3)
- 而`收藏`、`稍后再看`会显示所有分 P 的总时长：
  - `收藏`：
  ![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/45d0fb4b-aeb9-4fb8-aca3-00663a5b0b2a)
  - `稍后再看`：
  ![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/4c5e974b-a544-4bc0-957f-669264c36483)

### 另附赠 Android 客户端[^1]相关显示效果：

- `历史记录`只会显示最近观看分 P 的时长，且封面右下角会显示最近观看分 P 的具体观看进度（形如`具体看到的时间点/最近观看分 P 的时长`），封面下方也还有一进度条指示大致观看进度；并会在标题下方显示最近观看分 P 名（同时不显示视频所属 UP 主）：
![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/b8ff8b91-11ee-4f21-90cf-481b63d9a5ca)
- `收藏`只会显示分 P 数：
![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/2ecc01b4-40cf-4933-b62c-5c653f9d502e)
- `稍后再看`会显示所有分 P 的总时长，且封面右下角会显示所有分 P 的具体观看进度（形如`具体看到的时间点/所有分 P 的总时长`），封面下方也还有一进度条指示大致观看进度（说来有趣顺带一提，这里显示的弹幕数居然精确是显示到个位的）：
![image](https://github.com/the1812/Bilibili-Evolved/assets/49396421/540e610a-6927-41b5-83f3-7f13e25fa6be)

## 建议

1. `自定义顶栏`显示分 P 数相关

   - [ ] `历史`组件能显示出分 P 数（若为多 P 视频）
   - [ ] `收藏`组件能显示出分 P 数（若为多 P 视频）
   - [ ] `稍后再看`组件能显示出分 P 数（若为多 P 视频）

2. `自定义顶栏`显示时长相关：

   - [ ] 针对多 P 视频，统一`历史`和`稍后再看`组件中显示的视频时长类型（令人感叹的是现在新版原版顶栏也是不统一的）
     - [ ] 1. 个人建议再给个选项选择是默认显示最近观看分 P（没看过的视频就显示第 1 P）的时长还是所有分 P 的总时长
     - [ ] 2. 光标悬停可切换上述两种时长的显示

3. `自定义顶栏`显示观看进度相关：
   
   - [ ] 1. `稍后再看`组件也显示观看进度（之前在 #4346 提过一嘴）（别忘了封面下方的进度条）
   - [ ] 2. 给个选项选择观看进度的展现形式是百分比还是具体的观看进度（形如`具体看到的时间点/时长`）
   - [ ] 3. 针对多 P 视频，统一`历史`和`稍后再看`组件中显示的观看进度类型（最近观看分 P 的观看进度还是所有分 P 的观看进度）。
        显示的观看进度类型和显示的视频时长类型应该是同步的，跟随 2.1 的选项变化而变化，所以：
        - 若 3.2 选择的观看进度的展现形式为具体的观看进度，此项与上述时长应是一体的
        - 若 3.2 选择的观看进度的展现形式为百分比，此项与 #4346 的`已观看`应是一体的（即如现在`自定义顶栏`中的`历史`组件，100% 时就显示为`已看完`〔破站这文案不统一也是醉了〕）
   - [ ] 4. 针对多 P 视频，`历史`和`稍后再看`组件中显示最近看到哪一 P（个人建议形如`x/y`，没看过的视频就显示为`-/y`）
        或者可以做成光标悬停切换最近看到哪一 P 和分 P 数，这样此项和分 P 数就是一体的
     - [ ] 个人建议再给个选项选择是默认显示最近看到哪一 P 还是分 P 数

就是感觉依部分组件现在的设计可能放不下这么多信息，要是再考虑 #916 要加的删除按钮……

**Comment by Geeyun-JY3** [comment]
2024-06-04 编辑说明：

- 将`稍后再看`纳入功能建议适用范围
- 补充了旧版原版顶栏和`自定义顶栏`的相关截图，完善了相关现状说明
- 加入了新版原版顶栏的相关截图和现状说明
- 在上述基础上加入了现状省流表格方便对比
- 重新梳理了功能建议，在原先的功能建议基础上对其进行拆分、细化、扩展，使其更具条理性、可行性

**Comment by the1812** [comment]
看了一下三个组件的 API 中可以获取到的信息:

| API 包含信息      | 历史（记录） | 收藏 | 稍后再看 |
| ----------------- | ------------ | ---- | -------- |
| 分 P 数           | ✔            | ✔    | ✔        |
| 各分 P 时长       | ❌            | ❌    | ✔        |
| 总时长            | ❌            | ✔    | ✔        |
| 当前是第几 P      | ✔            | \    | ✔        |
| 当前分 P 时长     | ✔            | \    | ✔        |
| 当前分 P 观看进度 | ✔            | \    | ✔        |

> 目前稍后再看的 "当前分 P 观看进度" 计算方式也有问题, 使用了当前分 P 已看时长除以总时长.

- 三个组件显示出分 P 数, 可行.
- 若要统一 `历史` 和 `稍后再看` 组件中显示的视频时长类型, 就只能使用 `当前分 P 时长`, 观看进度也是如此.
- 关于观看进度的展现形式, 都支持封面下方的进度提示后, 文字应展示具体时间点, 百分比就不再需要了.
- ~~`历史` 中显示最近看到哪一 P, 无法实现.~~

**Reply by Geeyun-JY3** [reply]
`稍后再看`的观看进度目前好像只有封面下方的进度提示啊，有文字展示具体时间点吗？

**Reply by Geeyun-JY3** [reply]
而且`历史`目前好像还是不会显示分 P 数啊。

**Reply by the1812** [reply]
好像漏加了...

**Reply by Geeyun-JY3** [reply]
`稍后再看`的文字观看进度是不是也漏加了？

**Reply by the1812** [reply]
> 稍后再看的文字观看进度是不是也漏加了？

砍了, 挤不下了, 就放个进度条了

**Comment by Geeyun-JY3** [comment]
才发现网页端历史页面还会显示个`已收藏`标识……：
![image](https://github.com/user-attachments/assets/8307167c-58bb-493d-b16b-83fb0bff6429)
API 是不是拿不到这个信息？

**Reply by the1812** [reply]
拿得到, 不过历史记录为啥要关心这个?

**Reply by Geeyun-JY3** [reply]
我也很奇怪为什么历史记录页面会有这个标识……做不做进`自定义顶栏`我倒是无所谓。
哪天历史记录页面也改成新版 UI 这标识估计就没了……
